libWlz/WlzConvolve.3
'\" t
.\" ident MRC HGU $Id$
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" Project:    Woolz							"
.\" Title:      WlzConvolve.3			                      	"
.\" Date:       March 1999	                                    	"
.\" Author:     Bill Hill 				    		"
.\" Copyright:	1999 Medical Research Council, UK.			"
.\"		All rights reserved.					"
.\" Address:	MRC Human Genetics Unit,				"
.\"		Western General Hospital,				"
.\"		Edinburgh, EH4 2XU, UK.					"
.\" Purpose:    Woolz function which space domain convolutions.		"
.\" $Revision$
.\" Maintenance:Log changes below, with most recent at top of list.	"
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH "WlzConvolve" 3 "%G%" "MRC HGU Woolz" "Woolz Procedure Library"
.SH NAME
WlzConvolve \- Performs space-domain convolution.
.SH SYNOPSIS
.LP
.B #include <Wlz.h>
.LP
.BI "WlzObject *WlzConvolve(WlzObject *" "srcObj" ,
.BI "WlzConvolution *" "conv" ,
.ti +10m
.BI "int " "newObjFlag" ,
.BI "WlzErrorNum  *" wlzErr );
.LP
.BI "int WlzConvolveSeqParFn(WlzSeqParWSpace *" spWSpace ,
.BI "void *" "spData"
.ti +10m
.BI "WlzErrorNum  *" wlzErr );
.SH ARGUMENTS
.TS
tab(^);
lI l.
srcObj^source object pointer.
conv^convolution data structure
newObjFlag^new object or in\-place flag, non\-zero for a new object
spWSpace^WlzSeqPar workspace data structure
spData^used to pass \fIconv\fR through WlzSeqPar
wlzErr^destination error code pointer, may be null
.TE
.SH MT-LEVEL
.LP
Safe
.SH DESCRIPTION
WlzConvolve performs a space\-domain convolution on a
WLZ_2D_DOMAINOBJ object with integral (int, short or unsigned char)
grey values.
WlzConvolve returns a pointer to the convolved object or a NULL
pointer on error.
.SH EXAMPLES
.LP
.ps -2
.cs R 24
.nf
/*
 * Simple example of using WlzConvolve() to perform a space
 * domain 3x3 mean convolution filter (See WlzSeqPar(3)).
 */

#include <stdio.h>
#include <stdlib.h>
#include <Wlz.h>

int             main(char *argv[], int argc)
{
  WlzObject     *inObj = NULL,
                *outObj = NULL;
  WlzErrorNum   errNum = WLZ_ERR_NONE;
  WlzConvolution conv;
  const int     meanMask[]
  {
    1, 1, 1,
    1, 1, 1,
    1, 1, 1
  };

  conv.type = WLZ_CONVOLVE_INT;
  conv.linkcount = 0;
  conv.xsize = 3;
  conv.ysize = 3;
  conv.cv = meanMask;
  conv.divscale = 9;
  sConv.offset = 0;
  sConv.modflag = 0;
  if((inObj= WlzReadObj(stdin, &errNum)) == NULL)
  {
    (void )fprintf(stderr, "%s: Failed to read input object.\n", *argv);
  }
  else if((outObj = WlzConvolveObj(inObj, &conv, 1, &errNum)) == NULL) ||
          (errNum != WLZ_ERR_NONE))
  {
    (void )fprintf(stderr, "%s: Failed to convolve object.\n", *argv);
  }
  else if((errNum = WlzWriteObj(stdout, outObj)) != WLZ_ERR_NONE)
  {
    (void )fprintf(stderr, "%s: Failed to write output object.\n", *argv);
  }
  return(errNum);
}
.fi
.cs R
.ps +2
.SH SEE ALSO
WlzSeqPar(3), WlzError(3)
.SH BUGS
Still to be found.


ChangeLog
$Log$
Revision 1.2  2007/12/05 13:33:30  bill
*** empty log message ***



libWlz/WlzConvolve.c
#if defined(__GNUC__)
#ident "University of Edinburgh $Id$"
#else
static char _WlzConvolve_c[] = "University of Edinburgh $Id$";
#endif
/*!
* \file         libWlz/WlzConvolve.c
* \author       Jim Piper, Bill Hill
* \date         March 1999
* \version      $Id$
* \par
* Address:
*               MRC Human Genetics Unit,
*               MRC Institute of Genetics and Molecular Medicine,
*               University of Edinburgh,
*               Western General Hospital,
*               Edinburgh, EH4 2XU, UK.
* \par
* Copyright (C), [2012],
* The University Court of the University of Edinburgh,
* Old College, Edinburgh, UK.
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be
* useful but WITHOUT ANY WARRANTY; without even the implied
* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
* PURPOSE.  See the GNU General Public License for more
* details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the Free
* Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA  02110-1301, USA.
* \brief	Functions for convolving the values of domain objects.
* \ingroup	WlzValuesFilters
*/

#include <stdlib.h>
#include <Wlz.h>

/*!
* \return	Convolved pixel value.
* \ingroup	WlzValuesFilters
* \brief	Performs a general space-domain convolution for
*               WlzSeqPar().
* \param	spWSpace		Work space data structure from
* 					WlzSeqPar().
* \param	spData			Data passed on by WlzSeqPar(),
*					which is used to pass the convolution
*					structure.
*/
int 		WlzConvolveSeqParFn(WlzSeqParWSpace *spWSpace,
				    void *spData)
{
  WlzConvolution *conv;
  int		*data,
  		*mask;
  int		idX,
  		idY,
		lx,
		ly,
		convPixVal = 0;

  WLZ_DBG((WLZ_DBG_LVL_FN|WLZ_DBG_LVL_3),
  	  ("WlzConvolveSeqParFn FE %p %p\n",
	   spWSpace, spData));
  if(spWSpace && ((conv = (WlzConvolution *)spData) != NULL) && 
      conv->divscale)
  {
    mask = conv->cv;
    lx = (conv->xsize - 1) / 2; 
    ly = (conv->ysize - 1) / 2;
    WLZ_DBG((WLZ_DBG_LVL_3), 
	    ("WlzConvolveSeqParFn 01 %d %d\n",
	     lx, ly));
    for(idY = -ly; idY <= ly; ++idY)
    {
      data = spWSpace->adrptr[idY] - lx;
      for(idX = -lx; idX <= lx; ++idX)
      {
	WLZ_DBG((WLZ_DBG_LVL_3), 
		("WlzConvolveSeqParFn 02 %d %d\n",
		 *data, *mask));
	convPixVal += *data++ * *mask++;
      }
    }
    convPixVal = (convPixVal / conv->divscale) + conv->offset;
    if(conv->modflag)
    {
      convPixVal = abs(convPixVal);
    }
  }
  WLZ_DBG((WLZ_DBG_LVL_FN|WLZ_DBG_LVL_3),
  	  ("WlzConvolveSeqParFn FX %d\n",
	   convPixVal));
  return (convPixVal);
}

/*!
* \return	Convolved object or NULL on error.
* \ingroup	WlzValuesFilters
* \brief	Performs a general space-domain convolution using WlzSeqPar().
*		Only objects with WLZ_EMPTY_OBJ and WLZ_2D_DOMAINOBJ
*               types are valid. WLZ_2D_DOMAINOBJ ojects must have
*               non null domain and values fields, and only integral
*               values (ie int, short or WlzUByte) are valid.
* \param	inObj			Given object.
* \param	conv			Convolution data structure.
* \param	newObjFlag		If zero the convolution is done
*					in place, else a new object is created.
* \param	dstErr			Destination error pointer, may be NULL.
*/
WlzObject 	*WlzConvolveObj(WlzObject *inObj, WlzConvolution *conv,
			        int newObjFlag, WlzErrorNum *dstErr)
{
  int		bkgIntVal,
  		convSize;
  WlzObject 	*outObj = NULL;
  WlzErrorNum	errNum = WLZ_ERR_NONE;
  WlzPixelV	bkgVal;

  WLZ_DBG((WLZ_DBG_LVL_FN|WLZ_DBG_LVL_1),
  	  ("WlzConvolveObj FE %p %p %d\n",
	   inObj, conv, newObjFlag));
  if((inObj == NULL) || (conv == NULL))
  {
    errNum = WLZ_ERR_OBJECT_NULL;
  }
  else if(inObj->domain.core == NULL)
  {
    errNum = WLZ_ERR_DOMAIN_NULL;
  }
  else if(inObj->values.core == NULL)
  {
    errNum = WLZ_ERR_VALUES_NULL;
  }
  else if(conv->type != WLZ_CONVOLVE_INT)
  {
    errNum = WLZ_ERR_GREY_DATA;
  }
  else
  {
    switch(inObj->type)
    {
      case WLZ_EMPTY_OBJ:
	outObj = (newObjFlag)? WlzMakeEmpty(&errNum): inObj;
	break;
      case WLZ_2D_DOMAINOBJ:
	bkgVal = WlzGetBackground(inObj, &errNum);
	switch(bkgVal.type)
	{
	  case WLZ_GREY_INT:
	    bkgIntVal = bkgVal.v.inv;
	    break;
	  case WLZ_GREY_SHORT:
	    bkgIntVal = bkgVal.v.shv;
	    break;
	  case WLZ_GREY_UBYTE:
	    bkgIntVal = bkgVal.v.ubv;
	    break;
	  default:
	    errNum = WLZ_ERR_GREY_DATA;
	    break;
	}
	if(errNum == WLZ_ERR_NONE)
	{
	  convSize = (WLZ_MAX(conv->xsize, conv->ysize) - 1) / 2;
	  outObj = WlzSeqPar(inObj, newObjFlag, 0, WLZ_RASTERDIR_ILIC,
			     convSize, bkgIntVal,
			     (void *)conv, WlzConvolveSeqParFn, &errNum);
	}
	break;
      default:
        errNum = WLZ_ERR_OBJECT_TYPE;
	break;
    }
  }
  if( dstErr )
  {
    *dstErr = errNum;
  }
  WLZ_DBG((WLZ_DBG_LVL_FN|WLZ_DBG_LVL_1),
  	  ("WlzConvolveObj FX %p\n",
	   outObj));
  return(outObj);
}

/*!
* \return	Sum of convolution values.
* \ingroup	WlzValuesFilters
* \brief	Calculates the sum of the values in the convolution object.
* \param	conv			Given convolution object.
*/
int		WlzConvolutionSum(WlzConvolution *conv)
{
  int		count,
  		sum = 0;
  int		*values;

  if(conv && ((count = conv->xsize * conv->ysize) > 0) &&
     ((values = conv->cv) != NULL))
  {
    sum = *values;
    while(--count > 0)
    {
      sum += *++values;
    }
  }
  return(sum);
}

/*!
* \return	Convolution sum.
* \ingroup	WlzValuesFilters
* \brief	Normalises a convolution object by adjusting the scaling such
* 		that the components effectively sum to unity.
* \param	conv			Given convolution object.
*/
int		WlzConvolutionNormalise(WlzConvolution *conv)
{
  int		sum = 0;

  if(conv)
  {
    if((sum = WlzConvolutionSum(conv)) == 0)
    {
      conv->divscale = 1;
    }
    else
    {
      conv->divscale = sum;
    }
  }
  return(sum);
}
